packages/react-router-dom/index.tsx

```diff
@@ -993,30 +993,33 @@ export function useSearchParams(
 
   let defaultSearchParamsRef = React.useRef(createSearchParams(defaultInit));
   let hasSetSearchParamsRef = React.useRef(false);
+  let searchParamsRef = React.useRef<URLSearchParams>(defaultSearchParamsRef.current);
 
   let location = useLocation();
   let searchParams = React.useMemo(
-    () =>
+    () => {
       // Only merge in the defaults if we haven't yet called setSearchParams.
       // Once we call that we want those to take precedence, otherwise you can't
       // remove a param with setSearchParams({}) if it has an initial value
-      getSearchParamsForLocation(
+      searchParamsRef.current = getSearchParamsForLocation(
         location.search,
         hasSetSearchParamsRef.current ? null : defaultSearchParamsRef.current
-      ),
+      );
+      return searchParamsRef.current;
+    },
     [location.search]
```

Comment:
I'm fairly certain that while this works in most cases, it's not suspense-safe.

Quoting [the rules](https://react.dev/reference/react/useRef#avoiding-recreating-the-ref-contents) for `useRef`:
> Do not write or read ref.current during rendering, except for initialization. This makes your component’s behavior unpredictable.

Writing to `searchParamsRef` during render like this violates those rules, and I wouldn't be surprised if it leads to inconsistent states when using `startTransition`, which may be a concern for RRv7.

I think the correct solution in the future would be to wrap `setSearchParams` in `useEffectEvent`, but that isn't stable yet 🙃

Reply:
> I'm fairly certain that while this works in most cases, it's not transition-safe.
> 
> Quoting [the rules](https://react.dev/reference/react/useRef#avoiding-recreating-the-ref-contents) for `useRef`:
> 
> > Do not write or read ref.current during rendering, except for initialization. This makes your component’s behavior unpredictable.
> 
> Writing to `searchParamsRef` during render like this violates those rules, and I wouldn't be surprised if it leads to inconsistent states when using `startTransition`, which may be a concern for RRv7.
> 
> I think the correct solution in the future would be to wrap `setSearchParams` in `useEffectEvent`, but that isn't stable yet 🙃

In context, if you take a look at the example code in question, they are saying not to set the reference's value during the body of the functional component - or equivalently the render function on a class component. It is not saying that you should not assign references within the context of a useEffect (or useMemo specifically in this case).

Reply:
> What about this

I think that should be alright, would just need to validate that it works

> It is not saying that you should not assign references within the context of a useEffect (or useMemo specifically in this case)

You're right in that you can write to refs in `useEffect`, but you still shouldn't do it in `useMemo` since the callback of a `useMemo` does get ran during render

Reply:
> You're right in that you can write to refs in `useEffect`, but you still shouldn't do it in `useMemo` since the callback of a `useMemo` does get ran during render (if its dependencies change ofc)

`useMemo` executes between re-renders just like `useEffect`: https://react.dev/reference/react/useMemo

Reply:
> useMemo executes between re-renders just like useEffect
> All hooks behave like this; executing asynchronously from the rendering logic.

This isn't true. Effects are asynchronous, but memos are synchronous and are evaluated during render (if their dependencies change). This can be proved by looking at the `ReactFiberHooks.jsx` implementation of `useMemo`. Notice that the execution of `nextCreate` isn't deferred, rather it's called inline during render.

![image](https://github.com/remix-run/react-router/assets/14191578/b0efcf0a-e3e1-430b-8f64-6500cc366733)

> If hooks were just standard function calls that executed synchronously, then there wouldn't be any reason to dub them `hooks` and demand they follow special rules.

In the case of `useMemo`, the special thing that makes it a hook is the fact that it caches previous executions, as the cache is stored as part of the hook tree's state. The function you pass to a memo should have 0 side effects, so whether it's executed synchronously or asynchronously shouldn't matter.

Reply:
Honestly I don't think that setting refference during useMemo could cause any issues cos it is non-reactive type and it would need to hit same event loop tick in specific order (possibility close to zero), but since this is widely used public repo and it is close to zero, but not a zero, I would choose recommended solution just to avoid these long threads..

Reply:
> > useMemo executes between re-renders just like useEffect
> > All hooks behave like this; executing asynchronously from the rendering logic.
> 
> This isn't true. Effects are asynchronous, but memos are synchronous and are evaluated during render (if their dependencies change). This can be proved by looking at the `ReactFiberHooks.jsx` implementation of `useMemo`. Notice that the execution of `nextCreate` isn't deferred, rather it's called inline during render.z

Please see the official documentation, which states that `useMemo is a React Hook that lets you cache the result of a calculation between re-renders.` (https://react.dev/reference/react/useMemo). If their documentation is wrong then you should bring that up with the react team, but more likely your interpretation of the code is wrong.

Diving into the code, here is the file that defines the react hooks: https://github.com/facebook/react/blob/0cdfef19b96cc6202d48e0812b5069c286d12b04/packages/react/src/ReactHooks.js#L122

We see here that useEffect has the following definition:
```
export function useEffect(
  create: () => (() => void) | void,
  deps: Array<mixed> | void | null,
): void {
  const dispatcher = resolveDispatcher();
  return dispatcher.useEffect(create, deps);
}
```

And here is the definition for useMemo:
```
export function useMemo<T>(
  create: () => T,
  deps: Array<mixed> | void | null,
): T {
  const dispatcher = resolveDispatcher();
  return dispatcher.useMemo(create, deps);
}
```

Note that both of them are called via the dispatcher (asynchronous); the `create` function is not invoked synchronously. 

> Also, the 'between rerenders' part means that the value is cached between rerenders, not that the function is called between rerenders.

As seen in the definition above `useMemo` executes asynchronously via the dispatcher.

> > If hooks were just standard function calls that executed synchronously, then there wouldn't be any reason to dub them `hooks` and demand they follow special rules.
> 
> In the case of `useMemo`, the special thing that makes it a hook is the fact that it caches previous executions, as the cache is stored as part of the hook tree's state. The function you pass to a memo should have 0 side effects, so it shouldn't care how it's executed, but if memos themselves were asynchronous then you'd have all sorts of stale closure problems.

React repeatedly processes hooks so long as their are more changes, which is how it avoids things getting stale / out of sync. Here is some sample code to demonstrate that `useMemo` executes asynchronously from the rendering process:

```
export interface MemoTestProps {
  n: number;
}

export const MemoTest = (props: MemoTestProps) => {
  const renderCount = useRef<number>(0);
  const memoCount = useRef<number>(0);
  const [i, setIndex] = useState<number>(-1);
  const [i2, setIndex2] = useState<number>(-1);

  useEffect(() => {
    renderCount.current = 0;
    memoCount.current = 0;
    setIndex(1);
    setIndex2(1);
  }, [props.n]);

  const sum = useMemo(() => {
    memoCount.current += 1;
    return (i2 * (i2 + 1) / 2);
  }, [i2]);
  
  useEffect(() => {
    if(i < props.n) {
      setIndex(prev => prev + 1);
    }
  }, [i, props.n, setIndex]);

  useEffect(() => {
    if(i2 < i - 2) {
      setIndex2(prev => prev + 1);
    }
  }, [i, i2, setIndex2]);

  renderCount.current += 1;
  return (
    <div>
      <p><label>Render Count:</label> <span>{renderCount.current}</span></p>
      <p><label>Memo Count:</label> <span>{memoCount.current}</span></p>
      <p><label>{i2}th Triangle Number:</label> <span>{sum}</span></p>
    </div>
  );
};

```

If I then render `<MemoTest n={5} />` I get the following output:

![image](https://github.com/remix-run/react-router/assets/106355404/e7c5722e-28c8-4dd4-8a12-a7ff72ca1e71)

As you can see, the number of times that the memo is run does not match the number of times that the component is rendered. Thus `useMemo` is executing asynchronously from the rendering process.

Reply:
@Brendonovich I did some more testing, and it would appear that you are correct - `useMemo` does execute synchronously with the re-rendering process. My test code:

```
export interface MemoTestProps {
  n: number;
}

export const MemoTest = (props: MemoTestProps) => {
  const [i, setIndex] = useState<number>(0);
  const renderCount = useRef<number>(0);
  const memoCount = useRef<number>(0);

  renderCount.current += 1;

  console.log(`renderCount before useMemo: ${renderCount.current}, memoCount: ${memoCount.current}`);
  const cachedValue = useMemo(() => {
    memoCount.current += 1;
    console.log(`updated memoCount: ${memoCount.current}`);
    return i;
  }, [i]);
  console.log(`renderCount after useMemo: ${renderCount.current}, memoCount: ${memoCount.current}`);
  
  useEffect(() => {
    if (i < 10) {
      setIndex(prev => prev + 1);
    }
  }, [i, setIndex]);

  return (
    <div>
      <p><label>Render Count:</label> <span>{renderCount.current}</span></p>
      <p><label>Memo Count:</label> <span>{memoCount.current}</span></p>
      <p><label>Cached Value:</label> <span>{cachedValue}</span></p>
    </div>
  );
};
```

The `memoCount` can be observed to always update synchronously with the `renderCount` in the log output.

I will find some time to look back over this to see if there is a better solution then; although the present solution is still far better than the current broken behavior.

Reply:
> I don't think that setting reference during useMemo could cause any issues cos it is non-reactive type and it would need to hit same event loop tick in specific order (possibility close to zero)

Event loop ticks aren't the problem, writing to a ref during render is simply not good as not all results of renders are actually _used_, so just because a value is calculated during render doesn't mean the user will see that value, as during a Transition the results of that render may get thrown away by React, and yet if you put that value into a ref during render it'll stick around and be accessible in other places, even though that render was never used.

I think the `useEffect` solution should be fine.

Reply:
I've updated the PR so that it uses a useEffect to updated the reference.